webui/react/src/services/stream/stream.ts

```diff
@@ -0,0 +1,236 @@
+    const sync_id = this.#numSyncs.toString();
+
+    const wholeSub = this.#curSub ? { ...this.#curSub, ...newSub } : newSub;
+
+    const payload = reduce(
+      wholeSub,
+      (payload, ent) => {
+        if (!ent) return payload;
+        const k = ent.spec.id();
+        const curSpec = this.#curSub?.[k];
+        const keyCache = curSpec?.keyCache || new KeyCache();
+        newSpecWithCache[k] = { ...ent, keyCache };
+        payload['known'][k] = keyCache.known();
+        payload['subscribe'][k] = {
+          ...ent.spec.toWire(),
+          since: keyCache.maxSeq(),
+        };
+        return payload;
+      },
+      { known: {}, subscribe: {}, sync_id: sync_id } as Record<string, StreamContent>,
+    );
+
+    this.#curSub = newSpecWithCache;
+    this.#ws!.send(JSON.stringify(payload));
+    this.#syncSent = sync_id;
+  }
+
+  #processPending(): void {
+    while (this.#pendingMsg.length > 0) {
+      const msg = this.#pendingMsg.shift();
+      if (!msg) break;
+      if (msg['sync_id']) {
+        if (!msg['complete']) {
+          this.#syncStarted = msg['sync_id'];
+        } else {
+          const completedSpecId = map(this.#curSub, (spec) => spec?.id || '').filter((i) => !!i);
+          this.#isLoaded?.(completedSpecId);
+          this.#syncComplete = msg['sync_id'];
+        }
+      } else if (this.#syncSent !== this.#syncStarted) {
+        // Ignore all messages between when we send a new subscription and when the
+        // sync-start message for that subscription arrives.  These are the online
+        // updates for a subscription we no longer care about.
+      } else {
+        forEach(msg, (val, k) => {
+          if (k.includes('_deleted')) {
+            const stream_key = trimEnd(k, '_deleted') as Streamable;
+            const deleted_keys = decode_keys(val as string);
+
+            this.#curSub?.[stream_key]?.keyCache.delete_msg(deleted_keys);
+            this.#onDelete(stream_key, deleted_keys);
+          } else {
+            this.#curSub?.[StreamEntityMap[k]]?.keyCache.upsert([val.id], val.seq);
+            this.#onUpsert(msg);
+          }
+        });
+      }
+    }
+  }
+
+  #processSubscription() {
+    if (!this.#curSub && this.#subs.length === 0) return;
+
+    let spec: SubscriptionGroup | undefined;
+
+    if (!this.#syncSent) {
+      // The websocket just connected/reconnected
+      // Resend current subscription if current sync not completed
+      if (this.#curSub && (!this.#syncComplete || this.#syncStarted !== this.#syncComplete)) {
+        spec = this.#curSub;
+      } else {
+        spec = this.#subs.shift();
+      }
+      spec && this.#sendSpec(spec);
+      return;
+    }
+
+    if (this.#syncComplete === this.#syncSent) {
+      // For established connection, only send a new sub when current sub is completed
+      // Skip and move to next in case of duplication
+      while (this.#shouldSkip(spec) && this.#subs.length > 0) {
+        spec = this.#subs.shift();
+      }
```

Comment:
this is a readability nit.

It is odd that `#shouldSkip()` would accept `undefined` as a value.  As far as I can tell, the only reason why it does is because it needs to for this loop to work right.  That means the details of this loop are leaking into a function defined elsewhere.

Perhaps:
```typescript
// have we finished offline messages for the current subscription?
if (this.#syncComplete !== this.#syncSent) return;

while (const spec = this.#subs.shift()) {
  // is this subscription worth sending?
  if (this.#shouldSkip(spec)) continue;
  // we have a fresh spec to send
  this.#sendSpec(spec);
  return;
}
```

Note that you don't need a length check because the output of `.shift()` when you have nothing in the list is `undefined`, so the `while()` loop will exit automatically in that case.

Then `#shouldSkip()` does not need to handle the `undefined` case, which makes it easier to reason about too.